gateway/app/drivers/geofencing/subscriptions.py

```diff
@@ -0,0 +1,364 @@
+import logging
+from datetime import datetime, timezone
+from enum import Enum
+from http import HTTPStatus
+from typing import Optional
+
+import geopy  # type: ignore[import-untyped]
+import geopy.distance  # type: ignore[import-untyped]
+import httpx
+from pydantic import AnyHttpUrl, AnyUrl
+
+from app.drivers.nef_auth import NEFAuth
+from app.exceptions import ApiException
+from app.interfaces.geofencing_subscriptions import (
+    GeofencingSubscriptionInterface,
+    GeofencingSubscriptionNotFound,
+)
+from app.redis import get_redis
+from app.schemas.geofencing import (
+    AreaEntered,
+    AreaLeft,
+    CloudEvent,
+    GeographicalCoordinates,
+    MonitoringEventSubscription,
+    MonitoringType,
+    NotificationEventType,
+    Protocol,
+    Status,
+    Subscription,
+    SubscriptionEventType,
+    SubscriptionRequest,
+)
+from app.settings import settings
+
+LOG = logging.getLogger(__name__)
+
+_prefix_subscription = "geofencing"
+_prefix_last_state = "geofencing_state"
+_prefix_counter = "geofencing_counter"
+_prefix_queue = "geofencing_queue"
+
+
+class State(str, Enum):
+    INSIDE = "INSIDE"
+    OUTSIDE = "OUTSIDE"
+
+
+def _handle_post_error(res: httpx.Response) -> None:
+    raise ApiException(
+        status=HTTPStatus.INTERNAL_SERVER_ERROR,
+        message="Error comunicating with the core",
+        code="INTERNAL_SERVER_ERROR",
+    )
+
+
+class RedisGeofencingSubscriptionInterface(GeofencingSubscriptionInterface):
+    def __init__(self, nef_url: AnyHttpUrl, nef_auth: NEFAuth) -> None:
+        super().__init__()
+        self.httpx_client = httpx.AsyncClient(base_url=str(nef_url), auth=nef_auth)
+
+    async def clear_expired_subscriptions(self) -> None:
+        redis = get_redis()
+        keys = await redis.keys(f"{_prefix_subscription}:*")
+
+        for key in keys:
+            sub = await redis.get(key)
+            if sub is None:
+                continue
+
+            sub = Subscription.model_validate_json(sub)
+            if (
+                sub.config.subscriptionExpireTime is None
+                or datetime.now(sub.config.subscriptionExpireTime.tzinfo)
+                < sub.config.subscriptionExpireTime
+            ):
+                continue
+
+            await self.delete_subscription(sub.id)
```

Comment:
Why don't you just set a TTL in the redis store and avoid all this work? Also `keys` is a very slow operation.